backends/qualcomm/qnn_quantizer.py

```diff
@@ -733,9 +837,199 @@ def _annotate_upsample2d(self, gm: torch.fx.GraphModule) -> None:
                 _annotated=True,
             )
 
+    def _annotate_embedding(self, gm: torch.fx.GraphModule) -> None:
+        op_sources = [torch.nn.modules.sparse.Embedding, torch.nn.modules.Embedding]
+        qconfig = self._get_quant_config(op_sources)
+        embedding_partitions = get_source_partitions(gm.graph, op_sources)
+
+        embedding_partitions = list(itertools.chain(*embedding_partitions.values()))
+        for embedding_partition in embedding_partitions:
+            embedding_node = embedding_partition.output_nodes[0]
+
+            weight = embedding_node.args[0]
+
+            input_qspec_map = {}
+            input_qspec_map[weight] = qconfig.input_activation
+
+            embedding_node.meta[QUANT_ANNOTATION_KEY] = QuantizationAnnotation(
+                input_qspec_map=input_qspec_map,
+                output_qspec=SharedQuantizationSpec((weight, embedding_node)),
+                _annotated=True,
+            )
+
+    def _annotate_softmax(self, gm: torch.fx.GraphModule) -> None:
+        op_sources = [torch.nn.Softmax, F.softmax]
+        qconfig = self._get_quant_config(op_sources)
+        softmax_partitions = get_source_partitions(gm.graph, op_sources)
+
+        softmax_partitions = list(itertools.chain(*softmax_partitions.values()))
+        for softmax_partition in softmax_partitions:
+            softmax_node = softmax_partition.output_nodes[0]
+            input_node = softmax_node.args[0]
+
+            input_qspec_map = {}
+            input_qspec_map[input_node] = qconfig.input_activation
```

Comment:
Does qnn not require fixed quantizaiton parameters for softmax? Meaning it can do any scale and zero point? If so I am curious if the implementation dequantizes first or not

Reply:
There is not a requirement which is fixed quantization parameters for softmax from the Qualcomm AI Engine Direct Doc. I think that it can do with reasonable scale, you could refer to Qualcomm AI Engine Direct Doc to get the detail.